### PR TITLE
Chain Sync CLI Commands

### LINKED ISSUE #536: Chain sync CLI commands
**Task summary**

CLI match of #535 and will depend on it

AC
- [ ] sync-wait RPC method is implemented
- [ ] the below commands can be called from the CLI client

COMMANDS: (from `lotus sync -h` as reference)
- [ ] status: check sync status
- [ ] wait: Wait for sync to be complete
- [ ] mark-bad: Mark the given block as bad, will prevent syncing to a chain that contains it
- [ ] check-bad: check if the given block was marked bad, and for what reason

**Specification reference**

- 

**Other information and links**

### PR DESCRIPTION
**Pull request by connormullett**
**Summary of changes**
Changes introduced in this pull request:
- Adds Chain-Sync CLI commands to forest. Blocks can be marked as bad blocks to never be synced. Users can check bad blocks and why they were marked as bad. Users can also ping the current state of the syncing process with two different commands. 

**Reference issue to close (if applicable)**

Closes 
#536 

**Other information and links**

### CODE REVIEW

**Review by cryptoquick — CHANGES_REQUESTED** [review]
I just checked out `sync wait` and `sync status`, and they look really good!

I'm not sure I know how to use `sync check-bad`, or at least, there's no output from that, and I'm not sure why. However, I also acknowledge that this isn't going to work great while we can't sync to mainnet.

Just some other minor nits and comments. Good work on this one, the serialization stuff seemed especially tricky!

**Comment by connormullett** [comment]
> I'm not sure I know how to use `sync check-bad`, or at least, there's no output from that, and I'm not sure why. However, I also acknowledge that this isn't going to work great while we can't sync to mainnet.

Check bad needs a block to be marked as bad. Otherwise there are no blocks marked as bad. I'll add in a condition to display "No blocks bad" or something if nothing is returned

File: node/rpc-client/src/sync_ops.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    filecoin_rpc::sync_check_bad(params).await
+}
+
+pub async fn mark_bad(params: SyncMarkBadParams) -> Result<SyncMarkBadResult, JsonRpcError> {
+    filecoin_rpc::sync_mark_bad(params).await
+}
+
+pub async fn status(params: SyncStateParams) -> Result<SyncStateResult, JsonRpcError> {
+    filecoin_rpc::sync_state(params).await
+}
```
[the remark this answers is not in the document]
Comment (connormullett) [reply node/rpc-client/src/sync_ops.rs:19]: Good point, fixed.

File: forest/src/cli/sync_cmd.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                        tipset.epoch()
+                    } else {
+                        0
+                    };
+
+                    let base_height = if let Some(tipset) = base {
+                        tipset.epoch()
+                    } else {
+                        0
+                    };
```
[the remark this answers is not in the document]
Comment (connormullett) [reply forest/src/cli/sync_cmd.rs:68]: 👍

File: blockchain/chain_sync/src/sync_state.rs [hunk 3]
```diff
@@ -74,24 +95,46 @@ impl SyncState {
...
         self.stage
     }
 
-    /// Returns the current [Tipset] the
+    /// Returns the current [Tipset]
     pub fn target(&self) -> &Option<Arc<Tipset>> {
         &self.target
     }
 
+    /// Return an Arc clone of the Base [Tipset]
```
[the remark this answers is not in the document]
Comment (connormullett) [reply blockchain/chain_sync/src/sync_state.rs:113]: 👍

File: blockchain/chain_sync/src/sync_state.rs [hunk 4]
```diff
@@ -2,11 +2,12 @@
 // SPDX-License-Identifier: Apache-2.0, MIT
 
 use blocks::{tipset::tipset_json::TipsetJsonRef, Tipset};
+use chrono::{DateTime, Duration, Utc};
 use clock::ChainEpoch;
-use serde::{Serialize, Serializer};
+use serde::Deserializer;
+use serde::{Deserialize, Serialize, Serializer};
```
[the remark this answers is not in the document]
Comment (connormullett) [reply blockchain/chain_sync/src/sync_state.rs:8]: 👍

**Review by cryptoquick — APPROVED** [review]
Great work! Glad my suggestions were helpful.

File: forest/src/cli/sync_cmd.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                let cid: Cid = cid.parse().unwrap();
+                let response = sync_check_bad((CidJson(cid),))
+                    .await
+                    .map_err(handle_rpc_err)
+                    .unwrap();
+
+                if response.is_empty() {
+                    println!("Block \"{}\" is not marked as a bad block", cid);
+                } else {
+                    println!("response");
```
Comment (creativcoder) [inline forest/src/cli/sync_cmd.rs:133]: Feels like an incomplete message?

**Review by creativcoder — APPROVED** [review]
lgtm!

**Pull request merged**